### ISSUE TITLE
Brief connectivity breaks don't result in a retried download when `enableRetryOnNetworkGain == true`

### ISSUE THREAD

**Issue by dougkeen** [tonyofrancis/Fetch#133]
Repro steps in sample app:

1. Go to Single Download Demo and start the download
2. Once the download begins, go into airplane mode
3. Come back from airplane mode within 5 seconds
4. Instead of queueing and restarting, the download fails with a `CONNECTION_TIMED_OUT` error, and a retry never happens.

A retry only occurs if the break in connectivity is sustained for 5 or more seconds. This is due to this catch block in `FileDownloaderImpl.kt`:

```kotlin
        } catch (e: Exception) {
            if (!interrupted && !terminated) {
                logger.e("FileDownloader", e)
                var error = getErrorFromThrowable(e)
                error.throwable = e
                if (retryOnNetworkGain) {
                    try {
                        Thread.sleep(5000) // <-- Here's the 5 second wait
                    } catch (e: InterruptedException) {
                        logger.e("FileDownloader", e)
                    }
                    if (!networkInfoProvider.isNetworkAvailable) { // <-- The download delegate only retries if the ConnectivityManager reports no network connection at this distinct point in time, 5 seconds after the error
                        error = Error.NO_NETWORK_CONNECTION
                    }
                }
                downloadInfo.downloaded = downloaded
                downloadInfo.total = total
                downloadInfo.error = error
                if (!terminated) {
                    delegate?.onError(download = downloadInfo)
                }
            }
        }
```

In real world scenarios, it's very common to have drops in connectivity that last less than 5 seconds, such as changes between cell data networks and wifi, or switching wifi networks. The current 5 second wait logic is a poor fit for those scenarios.

**Comment by dougkeen** [comment]
I was able to slap together an ugly fix here: https://github.com/phound-inc/Fetch/commit/c2deadab682de99965b7dc92edb3d62aba8425fd

I can pull request that if you want, but I think the better solution would be to set up a broadcast receiver to get connectivity events, and perform disconnect detection + retries based on those events.

**Comment by tonyofrancis** [comment]
@dougkeen Thanks for brining up this issue. It was one of my concerns when implementing retry. The temporary fix would actually work perfectly! I did try implementing a broadcast receiver at one point and performed some tests, but the issue was that the Connectivity Manager at times is slow at reporting network changes. For example even when I disconnected from the network the connectivity manager would report still connected causing the download to timeout and then the retry would never happen. The issue become more apparent across different devices. From my test it takes an average of around 8 seconds for a device to reconnect.  I would be happy to accept your pull request as the fix.
